external-crates/move/crates/move-vm-runtime/src/interpreter/eval.rs

```diff
@@ -0,0 +1,1158 @@
+/// function represented by the frame. Control comes back to this function on return or
+/// on call. When that happens the frame is changes to a new one (call) or to the one
+/// at the top of the stack (return). If the call stack is empty execution is completed.
+pub fn run(
+    start_state: MachineState,
+    data_store: &impl DataStore,
+    gas_meter: &mut impl GasMeter,
+    extensions: &mut NativeContextExtensions,
+    loader: &Loader,
+) -> VMResult<Vec<Value>> {
+    let link_context = data_store.link_context();
+    let resolver = start_state.current_frame.resolver(link_context, loader);
+
+    let mut run_context = RunContext {
+        loader,
+        link_context,
+        defn_resolver: resolver,
+        extensions,
+    };
+
+    let mut state = start_state;
+
+    let mut buf = String::new();
+    if DEBUG_STEP_PRINT {
+        let _ = state.debug_print_stack_trace(&mut buf, loader);
+        println!("Call Frame:\n{:?}", state.current_frame);
+        println!("{buf}");
+        buf.clear();
+    }
+
+    // Run until we're done or we produce an error and bail.
+    while step(&mut state, &mut run_context, gas_meter)? != StepStatus::Done {
+        if DEBUG_STEP_PRINT {
+            println!("-------------------------------------");
+            println!("Call Frame:\n{:?}", state.current_frame);
+            let _ = state.debug_print_stack_trace(&mut buf, loader);
+            println!("{buf}");
+            buf.clear();
+        }
+        continue;
+    }
+
+    // When we are done, grab the operand stack as the return type.
+    let MachineState { operand_stack, .. } = state;
+    let RunContext { .. } = run_context;
+    Ok(operand_stack.value)
+}
+
+fn step(
+    state: &mut MachineState,
+    run_context: &mut RunContext,
+    gas_meter: &mut impl GasMeter,
+) -> VMResult<StepStatus> {
+    let fun_ref = state.current_frame.function();
+    let instructions = fun_ref.code();
+    let pc = state.current_frame.pc as usize;
+    assert!(pc <= instructions.len());
+    let instruction = &instructions[pc];
+
+    trace!(
+        fun_ref,
+        &state.current_frame.locals,
+        state.current_frame.pc,
+        instruction,
+        run_context.defn_resolver,
+        state
+    );
+
+    fail_point!("move_vm::interpreter_loop", |_| {
+        Err(state.set_location(
+            PartialVMError::new(StatusCode::VERIFIER_INVARIANT_VIOLATION)
+                .with_message("Injected move_vm::interpreter verifier failure".to_owned()),
+        ))
+    });
+
+    profile_open_instr!(gas_meter, format!("{:?}", instruction));
+    if DEBUG_STEP_PRINT {
+        println!("Instruction: {instruction:?}");
+    }
+    // These are split out because `PartialVMError` and `VMError` are different types. It's unclear
+    // why as the hold identical data, but if we could combine them, we could entirely inline
+    // `op_step` into this function.
```

Comment:
They don't hold identical information, the main different piece is the Location in which the error was raised (i.e. what Move module it occurred in). 

They used to be one type but we split them out into different types because we had so many bugs of the location not being set properly. The type just forces the location/module being set as the last step.

Reply:
We could also do something like `VMError::change_location_i_promise_i_know_what_im_doing` if it helps make this loop cleaner

Reply:
also as the typo man: "the hold identical.." -> "they hold identical..."

Reply:
It would let us combine the loops. Barring that, if we had `try`(https://doc.rust-lang.org/stable/unstable-book/language-features/try-blocks.html) we could also combine the loops.